rust/kernel/io_mem.rs

```diff
@@ -0,0 +1,200 @@
+            Ok(unsafe { concat_idents!(rust_helper_, $name)(ptr as _) })
+        }
+    };
+}
+
+macro_rules! define_write {
+    ($name:ident, $try_name:ident, $type_name:ty) => {
+        /// Writes IO data to the given offset, known at compile time.
+        ///
+        /// If the offset is not known at compile time, the build will fail.
+        pub fn $name(&self, value: $type_name, offset: usize) {
+            Self::check_offset::<$type_name>(offset);
+            let ptr = self.ptr.wrapping_add(offset);
+            // SAFETY: The type invariants guarantee that `ptr` is a valid pointer. The check above
+            // guarantees that the code won't link if `offset` makes the write go out of bounds
+            // (including the type size).
+            unsafe { concat_idents!(rust_helper_, $name)(value, ptr as _) }
+        }
+
+        /// Writes IO data to the given offset.
+        ///
+        /// It fails if/when the offset (plus the type size) is out of bounds.
+        pub fn $try_name(&self, value: $type_name, offset: usize) -> Result {
+            if !Self::offset_ok::<$type_name>(offset) {
+                return Err(Error::EINVAL);
+            }
+            let ptr = self.ptr.wrapping_add(offset);
+            // SAFETY: The type invariants guarantee that `ptr` is a valid pointer. The check above
+            // returns an error if `offset` would make the write go out of bounds (including the
+            // type size).
+            unsafe { concat_idents!(rust_helper_, $name)(value, ptr as _) };
+            Ok(())
+        }
+    };
+}
+
+impl<const SIZE: usize> IoMem<SIZE> {
+    /// Tries to create a new instance of a memory block.
+    ///
+    /// The resource described by `res` is mapped into CPU's address space so that it can be
+    /// accessed directly. It is also consumed by this function so that it can't be mapped again
+    /// to a different address.
+    ///
+    /// # Safety
+    ///
+    /// Callers must ensure that either (a) the resulting interface cannot be used to initiate DMA
+    /// operations, or (b) that DMA operations initiated via the returned interface use DMA handles
+    /// allocated through the `dma` module.
+    pub unsafe fn try_new(res: Resource) -> Result<Self> {
+        // Check that the resource has at least `SIZE` bytes in it.
+        if res.size < SIZE.try_into()? {
+            return Err(Error::EINVAL);
+        }
+
+        // Try to map the resource.
+        // SAFETY: Just mapping the memory range.
+        let addr = unsafe { rust_helper_ioremap(res.offset, res.size) };
+        if addr.is_null() {
+            Err(Error::ENOMEM)
+        } else {
+            // INVARIANT: `addr` is non-null and was returned by `ioremap`, so it is valid.
+            Ok(Self { ptr: addr as usize })
+        }
+    }
+
+    const fn offset_ok<T>(offset: usize) -> bool {
+        if let Some(end) = offset.checked_add(core::mem::size_of::<T>()) {
+            if end <= SIZE {
```

Comment:
I'm not 100% sure, but those `readX()` and `writeX()` seems to require the address being naturally aligned (i.e. aligned to the size of read and write), find something in the doc: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/driver-api/device-io.rst#n327

Won't hurt to get this clear.

Reply:
Are you proposing that we check for natural alignment?

If it isn't a requirement for all platforms, I don't think we can do it here.

Reply:
> Are you proposing that we check for natural alignment?
> 

Yes, that's what's in my mind.

> If it isn't a requirement for all platforms, I don't think we can do it here.

I understand your point, but without the natural alignment checks, the Rust version `readX()` and `writeX()` seems `unsafe` to me.

Reply:
> > If it isn't a requirement for all platforms, I don't think we can do it here.
> 
> I understand your point, but without the natural alignment checks, the Rust version `readX()` and `writeX()` seems `unsafe` to me.

Except that `readb` and `writeb` are `safe`.

Reply:
> If it isn't a requirement for all platforms, I don't think we can do it here.

If it's a requirement for any platform, the driver shouldn't depend on it. Perhaps we should offer an `_unchecked` variant, and who uses it need to prove that they won't ever be ran on platforms without misalignment memory access support.

Reply:
The document linked clearly states that "on many platforms" this is a requirements. Not all.

But I do like the idea of offering two versions. I'll make this one this the checked one, and we can add an unchecked one in the future if/when we find a case where we need this.

Reply:
Note that the link is the only evidence I found talking about the alignment requirement, and actually it was added just recently in this [patch](https://lore.kernel.org/lkml/20210402090542.131194-10-marcan@marcan.st/), I don't find many reviews on the doc. `readX()` and `writeX()` are heavily used APIs in kernel, I'd expect there are a few places (e.g. comments or docs) saying the requirement, but I only find one.

Reply:
FWIW, send an [email](https://lore.kernel.org/linux-arch/YQQr+twAYHk2jXs6@boqun-archlinux/T/#u) to linux-arch for some clarification.

Reply:
> If it's a requirement for any platform, the driver shouldn't depend on it. Perhaps we should offer an `_unchecked` variant, and who uses it need to prove that they won't ever be ran on platforms without misalignment memory access support.

We could provide a safe one that only exists in the platforms where it is so -- that way people can freely use it for platform-dependent drivers, and if someone ever tries to copy-paste the code or port the module to another platform, they would get an error.

Reply:
What we have now: a version with an unadorned name that checks alignment -- it is safe (in that it doesn't require an `unsafe` block and doesn't issue unaligned reads/writes) for anyone to use, but they'll get an error if they try some unaligned operation.

If we ever find a device whose driver requires unaligned access, we can add an unsafe version (whose name will be suffixed with `_unchecked`) that doesn't have the alignment check, and whose callers must be convinced that it is safe.